db/src/openai_utils.rs

```diff
@@ -0,0 +1,38 @@
+use color_eyre::Result;
+use openai::chat::{ChatCompletionBuilder, ChatCompletionMessage, ChatCompletionMessageRole};
+use std::env;
+
+pub async fn generate_summary(content: &str) -> Result<String> {
+    let api_key = env::var("OPEN_AI_API_KEY").expect("OPEN_AI_API_KEY must be set");
+    openai::set_key(api_key);
+
+    let messages = vec![
+        ChatCompletionMessage {
+            role: ChatCompletionMessageRole::System,
+            content: Some("You are a helpful assistant.".to_string()),
+            name: None,
+            function_call: None,
+        },
+        ChatCompletionMessage {
+            role: ChatCompletionMessageRole::User,
+            content: Some(format!("Summarize the following article and make sure to highlight the important parts: {content}")),
+            name: None,
+            function_call: None,
+        },
+    ];
+    let request = ChatCompletionBuilder::default()
+        .model("gpt-4o".to_string())
+        .messages(messages)
+        .max_tokens(4096u64)
+        .temperature(0.7)
+        .top_p(1.0)
```

Comment:
The `max_tokens` and `temperature` I'm familiar with, what is `top_p`/what does it control?

Reply:
The .top_p parameter in the context of the OpenAI API is used to control nucleus sampling, which is a technique for generating text. 

reference: https://community.openai.com/t/temperature-top-p-and-top-k-for-chatbot-responses/295542

It's basically letting the model consider all smallest amount of possible tokens, the model considers only the smallest set of tokens whose cumulative probability is greater than or equal to the top_p value

in my code I have it `top_p(1.0)` which means the model will consider all possible tokens which doesn't really make sense to have it, so I might remove it 😃